### PR TITLE
Augmented attributes interface-ref and metric fields to mpls-static lsp

### PR DESCRIPTION
**Pull request by Shashank-arista**
* (M) release/models/mpls/openconfig-mpls-static.yang
* (M) release/models/mpls/openconfig-mpls-types.yang
* (M) release/models/mpls/openconfig-mpls.yang
* (M) release/models/mpls/openconfig-mpls-igp.yang
* (M) release/models/mpls/openconfig-mpls-te.yang

### Change Scope

* Augmented with below attributes interface-ref, and metric in the path /network-instances/network-instance/mpls/lsps/static-lsps/static-lsp
```
  +--rw network-instances
     +--rw network-instance [name]
        +--rw mpls
           +--rw lsps
              +--rw static-lsps
                 +--rw static-lsp [name]
                    +--rw ingress
                    |  +--rw config
                    |  |  +--rw interface?        -> /oc-if:interfaces/interface/name
                    |  |  +--rw subinterface?     -> /oc-if:interfaces/interface[oc-if:name=current()/../interface]/subinterfaces/subinterface/index
                    |  |  +--rw metric?           uint8
                    |  +--ro state
                    |     +--ro interface?        -> /oc-if:interfaces/interface/name
                    |     +--ro subinterface?     -> /oc-if:interfaces/interface[oc-if:name=current()/../interface]/subinterfaces/subinterface/index
                    |     +--ro metric?           uint8
                    +--rw transit
                    |  +--rw config
                    |  |  +--rw interface?        -> /oc-if:interfaces/interface/name
                    |  |  +--rw subinterface?     -> /oc-if:interfaces/interface[oc-if:name=current()/../interface]/subinterfaces/subinterface/index
                    |  |  +--rw metric?           uint8
                    |  +--ro state
                    |     +--ro interface?        -> /oc-if:interfaces/interface/name
                    |     +--ro subinterface?     -> /oc-if:interfaces/interface[oc-if:name=current()/../interface]/subinterfaces/subinterface/index
                    |     +--ro metric?           uint8
                    +--rw egress
                       +--rw config
                       |  +--rw interface?        -> /oc-if:interfaces/interface/name
                       |  +--rw subinterface?     -> /oc-if:interfaces/interface[oc-if:name=current()/../interface]/subinterfaces/subinterface/index
                       |  +--rw metric?           uint8
                       +--ro state
                          +--ro interface?        -> /oc-if:interfaces/interface/name
                          +--ro subinterface?     -> /oc-if:interfaces/interface[oc-if:name=current()/../interface]/subinterfaces/subinterface/index
                          +--ro metric?           uint8 
```
### Platform Implementations

 * Arista: https://www.arista.com/en/um-eos/eos-mpls-commands#xx1141474
 CLI configuration: mpls static top-label top_tag[ bgp peer [peer IP]] [DEST_INTF] NEXTHOP_ADDR ACTION [PRIORITY]

### CODE REVIEW

**Review by dplore — CHANGES_REQUESTED** [review]
Please add some additional description where pointed out.  I will help with an additional implementation reference.

File: release/models/mpls/openconfig-mpls-types.yang [hunk 1]
```diff
@@ -126,6 +133,30 @@ module openconfig-mpls-types {
...
+  identity PAYLOAD_TYPE {
+    description
+      "base identity for supported MPLS
+       payload types";
+  }
+
+  identity PAYLOAD_MPLS {
+    base "PAYLOAD_TYPE";
+    description
+      "payload type MPLS";
```
[the remark this answers is not in the document]
Comment (Shashank-arista) [reply release/models/mpls/openconfig-mpls-types.yang]: I have updated the description as "Underlying payload type of the MPLS packet". Please check.

File: release/models/mpls/openconfig-mpls-static.yang [hunk 2]
```diff
@@ -170,6 +178,23 @@ submodule openconfig-mpls-static {
...
+
+    // interface-ref
+    uses oc-if:interface-ref-common;
+
+    leaf payload-type {
+      type identityref {
+        base oc-mplst:PAYLOAD_TYPE;
+      }
+      description
+        "MPLS payload type";
```
[the remark this answers is not in the document]
Comment (Shashank-arista) [reply release/models/mpls/openconfig-mpls-static.yang]: I have updated the description as "Specifies underlying payload type as MPLS label i.e., represents stacked MPLS labels". Please check

File: release/models/mpls/openconfig-mpls-static.yang [hunk 3]
```diff
@@ -170,6 +178,23 @@ submodule openconfig-mpls-static {
...
+    }
+
+    leaf metric {
+      type uint8;
+      description
+        "MPLS static lsp route metric";
```
[the remark this answers is not in the document]
Comment (Shashank-arista) [reply release/models/mpls/openconfig-mpls-static.yang]: Updated the description as "Specifies metric value used for the MPLS route". Please check

**Comment by dplore** [comment]
@Shashank-arista ping for any comments, especially regarding payload type.

**Comment by Shashank-arista** [comment]
> @Shashank-arista ping for any comments, especially regarding payload type.

Hi @dplore, Sorry for the delay in getting back on this. I tried to investigate more on payload-type requirement. Arista CLI needs a payload-type mandatorily, although today most of hardware platforms depend on auto-decide for inferring it. But still, there is a use case where payload-type would be necessary/used on a certain platform where we would need to skip applying egress ACLs for certain labels which relies on payload-type for the functionality.

[no] mpls static top-label <inlabel> <nhaddr> [ intf ] { { swap-label <outlabel> } | { pop [ payload-type { ipv4 | ipv6 } [ **access-list bypass** ] ]  } } [ metric <metricval> ]

So, it seems payload-type is something not a mandatory attribute but would still be required in certain specific scenario.

**Comment by dplore** [comment]
> > @Shashank-arista ping for any comments, especially regarding payload type.
> 
> Hi @dplore, Sorry for the delay in getting back on this. I tried to investigate more on payload-type requirement. Arista CLI needs a payload-type mandatorily, although today most of hardware platforms depend on auto-decide for inferring it. But still, there is a use case where payload-type would be necessary/used on a certain platform where we would need to skip applying egress ACLs for certain labels which relies on payload-type for the functionality.
> 
> [no] mpls static top-label [ intf ] { { swap-label } | { pop [ payload-type { ipv4 | ipv6 } [ **access-list bypass** ] ] } } [ metric ]
> 
> So, it seems payload-type is something not a mandatory attribute but would still be required in certain specific scenario.

Ok.  In this scenario since we haven't identified at least 2 vendors that support payload type, nor the intent to add a payload type, nor specific operational use cases I suggest the payload type be removed from the OC model.  It could added via an Arista specific yang augment.

**Comment by Shashank-arista** [comment]
> > > @Shashank-arista ping for any comments, especially regarding payload type.
> > 
> > 
> > Hi @dplore, Sorry for the delay in getting back on this. I tried to investigate more on payload-type requirement. Arista CLI needs a payload-type mandatorily, although today most of hardware platforms depend on auto-decide for inferring it. But still, there is a use case where payload-type would be necessary/used on a certain platform where we would need to skip applying egress ACLs for certain labels which relies on payload-type for the functionality.
> > [no] mpls static top-label [ intf ] { { swap-label } | { pop [ payload-type { ipv4 | ipv6 } [ **access-list bypass** ] ] } } [ metric ]
> > So, it seems payload-type is something not a mandatory attribute but would still be required in certain specific scenario.
> 
> Ok. In this scenario since we haven't identified at least 2 vendors that support payload type, nor the intent to add a payload type, nor specific operational use cases I suggest the payload type be removed from the OC model. It could added via an Arista specific yang augment.

Sure, @dplore. I have updated the pull request by removing the payload type related changes from OC model. Please review, thanks.

**Review by dplore — APPROVED** [review]
This LGTM.    Last call for comments!  I plan to merge this on April 17th, 2024.

**Pull request merged**